pkg/sethandler/controller.go

```diff
@@ -0,0 +1,227 @@
+	}
+	//If the Pod is not running, its containers haven't been created yet - no cpuset cgroup is present to be adjusted by the CPUSetter
+	if pod.Status.Phase != "Running" {
+		return false
+	}
+	return true
+}
+
+func (setHandler *SetHandler) determineCorrectCpuset(pod v1.Pod, container v1.Container) (types.Pool, error) {
+	for resourceName := range container.Resources.Requests {
+		resNameAsString := string(resourceName)
+		if strings.Contains(resNameAsString, resourceBaseName) && strings.Contains(resNameAsString, "shared") {
+			return setHandler.poolConfig.SelectPool(resNameAsString), nil
+		} else if strings.Contains(resNameAsString, resourceBaseName) && strings.Contains(resNameAsString, "exclusive") {
+			return setHandler.getListOfAllocatedExclusiveCpus(resNameAsString, pod, container)
+		}
+	}
+	return setHandler.poolConfig.SelectPool(resourceBaseName + "/default"), nil
+}
+
+func (setHandler *SetHandler) getListOfAllocatedExclusiveCpus(exclusivePoolName string, pod v1.Pod, container v1.Container) (types.Pool, error) {
+	checkpoint, err := checkpoint.GetCheckpoint()
+	if err != nil {
+		return types.Pool{}, errors.New("Kubelet checkpoint file could not be accessed because:" + err.Error())
+	}
+	podIDStr := string(pod.ObjectMeta.UID)
+	devices, err := checkpoint.GetComputeDeviceMap(podIDStr)
+	if err != nil {
+		return types.Pool{}, errors.New("List of assigned devices could not be read from checkpoint file for Pod: " + podIDStr + " because:" + err.Error())
+	}
+	exclusiveCpus, exist := devices[exclusivePoolName]
+	if !exist {
+		log.Println("WARNING: Container: " + container.Name + " in Pod: " + podIDStr + " asked for exclusive CPUs, but were not allocated any! Cannot adjust its default cpuset")
+		return types.Pool{}, nil
+	}
+	return calculateFinalExclusiveSet(exclusiveCpus, pod, container)
+}
+
+func calculateFinalExclusiveSet(exclusiveCpus *multusTypes.ResourceInfo, pod v1.Pod, container v1.Container) (types.Pool, error) {
+	var finalCPUSet types.Pool
+	var doesSetContainPinnerCore bool
+	for _, deviceID := range exclusiveCpus.DeviceIDs {
+		finalCPUSet.CPUs = finalCPUSet.CPUs + deviceID + ","
+		if deviceID == dedicatedPinnerCoreID {
+			doesSetContainPinnerCore = true
+		}
+	}
+	if isPinnerUsedByContainer(pod, container.Name) && !doesSetContainPinnerCore {
+		//1: Container is asking exclusive CPUs + 2: Container has a pool configured in the annotation = Pinner will be used
+		//If pinner is used by a container, we need to the add the configured CPU core holding its thread to their cpuset
+		finalCPUSet.CPUs = finalCPUSet.CPUs + dedicatedPinnerCoreID + ","
+	}
+	finalCPUSet.CPUs = strings.TrimSuffix(finalCPUSet.CPUs, ",")
+	return finalCPUSet, nil
+}
+
+func isPinnerUsedByContainer(pod v1.Pod, containerName string) bool {
+	for key, value := range pod.ObjectMeta.Annotations {
+		if strings.Contains(key, processConfigKey) {
+			var processConfig types.CPUAnnotation
+			err := json.Unmarshal([]byte(value), &processConfig)
```

Comment:
CPUAnnotation.Decode() could be used here. It makes some additional check for the annotation. The webhook does it also but as these are separate components, the checks could be done independently.

Reply:
While Webhook is indeed a separate component, it is guaranteed to validate the Pod before CPUSetter gets triggered.
CPUSetter watches changes in the API, but the Pod object won't be committed until Webhook patched, or rejected it, therefore at the time CPUSetter gets the notification, we can be sure that everything is syntactically correct.

So that's why I didn't want to repeat the same checks, I just assume the annotation is already validated, and I'm only interested in whether pinner will be used, or not.

Reply:
That is fine.